<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Foundational lesson for Dimensionality Reduction Module (structure is work in progress)">

<title>Introduction to Basis Vectors</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="basis_vectors_files/libs/clipboard/clipboard.min.js"></script>
<script src="basis_vectors_files/libs/quarto-html/quarto.js"></script>
<script src="basis_vectors_files/libs/quarto-html/popper.min.js"></script>
<script src="basis_vectors_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="basis_vectors_files/libs/quarto-html/anchor.min.js"></script>
<link href="basis_vectors_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="basis_vectors_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="basis_vectors_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="basis_vectors_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="basis_vectors_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction to Basis Vectors</h1>
</div>

<div>
  <div class="description">
    Foundational lesson for Dimensionality Reduction Module (structure is work in progress)
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-note callout-titled" title="Note to self (dcarey)">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note to self (dcarey)
</div>
</div>
<div class="callout-body-container callout-body">
<p>Existing resources: * PSA lecture - Representing things with data * Dimensionality reduction in strength testing.pptx * SpatialDimReduction.pptx * BasicsBasis.pptx * https://danielroelfs.com/blog/a-basic-comparison-between-factor-analysis-pca-and-ica/</p>
<p>MixOmics: This R package is handy and covers (s)PCA, (s)PLS-DA, CCA. It is designed for fusion of several omic layers (e.g., genomics and metabolomics). eBook attached.</p>
<p>Multiblock: This R packages covers multiblock data fusion. It has good RM or LiMM-SCA function. However, they are a bit hard to interpret. eBook attached and worth checking the vignettes.</p>
<p>ropls: Implementation of SIMCA in R. Mainly used for orthogonal PLS-DA (OPLS-DA). Check the vignette.</p>
<p>ALASCA: Good implementation of RM_ASCA functions. Visualisations are good too. Paper attached.</p>
</div>
</div>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This is a template tutorial.</p>
</section>
<section id="motivations" class="level2">
<h2 class="anchored" data-anchor-id="motivations">Motivations</h2>
<p>Why is dimensionality reduction important to learn about.</p>
<ul>
<li>What is the <em>“dimensionality”</em> of a dataset?</li>
<li>What does it mean to capture the variance of a dataset in fewer dimensions?</li>
</ul>
</section>
<section id="background-knowledge" class="level2">
<h2 class="anchored" data-anchor-id="background-knowledge">Background knowledge</h2>
<section id="datasets-as-matrices" class="level3">
<h3 class="anchored" data-anchor-id="datasets-as-matrices">Datasets as matrices</h3>
<p>A dataset can be thought of as a matrix because it’s essentially a two-dimensional arrangement of numbers, where rows represent individual observations or samples, and columns represent different variables or features.</p>
<p>This</p>
<p><strong>IMAGE: Excel spreadsheet to a matrix. Showing how some things like metadata and id columns get dropped. The “spreadsheet” is the data + organising and identifying metadata. The data matrix is the arrangements of the numbers alone.</strong></p>
<p>This matrix representation makes it easy to perform various mathematical operations and analyses on the dataset, such as matrix multiplication, linear algebra operations, and statistical computations. It’s a fundamental way of organizing and analyzing data, especially in fields like machine learning and data science.</p>
<section id="non-tabular-data" class="level4">
<h4 class="anchored" data-anchor-id="non-tabular-data">Non-tabular data</h4>
<p>A set of images can be treated as a data matrix by first representing each image as a vector, and then combining these vectors into a matrix. Here’s a general process:</p>
<p>Flatten the Images: Convert each image into a one-dimensional vector by concatenating its rows or columns. For example, if you have an image with dimensions 100x100 pixels, you would concatenate all the pixel values along either the rows (resulting in a vector of length 10000) or the columns.</p>
<p>Arrange the Vectors: Arrange these flattened vectors into rows or columns of a matrix. Each row or column of the matrix corresponds to one image in your dataset.</p>
<p><strong>IMAGE: images to a matrix</strong></p>
<p>A set of waveforms can be treated as a data matrix by representing each waveform as a vector and then combining these vectors into a matrix.</p>
<p><strong>IMAGE: biomech waveforms to a matrix</strong></p>
<p>Once represented as matrices, these data types can be manipulated, analyzed, and processed using various mathematical and computational techniques, including linear algebra, signal processing, and machine learning algorithms. This matrix representation allows for efficient storage, computation, and analysis of complex data structures.</p>
</section>
</section>
<section id="dimensionality" class="level3">
<h3 class="anchored" data-anchor-id="dimensionality">Dimensionality</h3>
<p>The dimensionality of a dataset refers to the number of features or variables used to describe each data point within that dataset. In simpler terms, it’s the number of columns or dimensions in your dataset.</p>
</section>
<section id="basis-vectors" class="level3">
<h3 class="anchored" data-anchor-id="basis-vectors">Basis vectors</h3>
<p>In math lingo, we can say that a basis vector is like a “building block” for describing points in space. You can combine these basis vectors in different amounts to reach any point in your space. They give you a way to break down complex positions into simpler parts.</p>
<section id="change-of-basis" class="level4">
<h4 class="anchored" data-anchor-id="change-of-basis">Change of basis</h4>
<p>In mathematics, particularly linear algebra, a change of basis operation is a transformation that allows you to express vectors in a new coordinate system. It’s like changing the perspective from which you view the vectors, while keeping their underlying meaning intact.</p>
</section>
</section>
</section>
<section id="example" class="level2">
<h2 class="anchored" data-anchor-id="example">Example:</h2>
<p>Let’s now work through a conrete (albeit contrived) example of dimensionality reduction in action to consolidate these background topics.</p>
<p><strong>2x simulated datasets both with two dimensions. One is made up sprint and hurdle times near perfect correlation, the other is sprint times and fingernail length (or something else obviously unrelated) with no correlation. Run a PCA on both and show some graphs. Do the calculations explicitly to show the basis transformation preserves the original values, and how “variance explained” is calculated.</strong></p>
</section>
<section id="cautions" class="level2">
<h2 class="anchored" data-anchor-id="cautions">Cautions</h2>
<p>No analysis method is without potential issues that can lead to misinterpretting data. Dimensionality reduction methods are no exception to this. As always there is no substitute for careful thinking, checking and re-checking, and being cautious.</p>
<section id="limitations-of-pca" class="level3">
<h3 class="anchored" data-anchor-id="limitations-of-pca">Limitations of PCA</h3>
<p>This artcicle <span class="citation" data-cites="dyer2023simplest">(<a href="#ref-dyer2023simplest" role="doc-biblioref">Dyer and Kording 2023</a>)</span> shows how PCA will return the same result on three very different datasets (reminiscent of the famous Anscombe’s quartet result). They argue that relaxing some of the constraints of PCA and using more flexible models may be beneficial in some situations.</p>
<p><strong>Check this against biomech data</strong>: Shinn <span class="citation" data-cites="shinn2023phantom">(<a href="#ref-shinn2023phantom" role="doc-biblioref">Shinn 2023</a>)</span> - “Here, we show that two common properties of data violate these assumptions (of PCA) and cause oscillatory principal components: smoothness, and shifts in time or space. These two properties implicate almost all neuroscience data”</p>
</section>
</section>
<section id="extensions" class="level2">
<h2 class="anchored" data-anchor-id="extensions">Extensions</h2>
<p>PCA of waveforms and functional PCA: A primer for biomechanics - <a href="https://github.com/johnwarmenhoven/PCA-FPCA" class="uri">https://github.com/johnwarmenhoven/PCA-FPCA</a></p>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-dyer2023simplest" class="csl-entry" role="listitem">
Dyer, Eva L, and Konrad Kording. 2023. <span>“Why the Simplest Explanation Isn’t Always the Best.”</span> <em>Proceedings of the National Academy of Sciences</em> 120 (52): e2319169120.
</div>
<div id="ref-shinn2023phantom" class="csl-entry" role="listitem">
Shinn, Maxwell. 2023. <span>“Phantom Oscillations in Principal Component Analysis.”</span> <em>bioRxiv</em>, 2023–06.
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>